src/sexp/list.rs

```diff
@@ -135,33 +141,51 @@ impl OwnedListSexp {
             )));
         }
 
-        let v: Sexp = v.into();
-
-        unsafe {
-            SET_VECTOR_ELT(self.values.inner(), i as _, v.0);
-        }
+        unsafe { self.set_value_unchecked(i, v.into().0) };
 
         Ok(())
     }
 
-    pub fn set_name(&mut self, i: usize, k: &str) -> crate::error::Result<()> {
-        // OwnedStringSexp::set_elt() checks the length, so don't check here.
+    /// # Safety
+    ///
+    /// The user has to assure bounds are checked.
+    #[inline]
+    pub unsafe fn set_value_unchecked(&mut self, i: usize, v: SEXP) {
+        unsafe { SET_VECTOR_ELT(self.values.inner(), i as _, v) };
+    }
 
-        if let Some(names) = self.names.as_mut() {
-            names.set_elt(i, k)?;
+    pub fn set_name(&mut self, i: usize, k: &str) -> crate::error::Result<()> {
+        if i >= self.len {
+            return Err(crate::error::Error::new(&format!(
+                "index out of bounds: the length is {} but the index is {}",
+                self.len, i
+            )));
         }
 
+        unsafe { self.set_name_unchecked(i, str_to_charsxp(k)?) };
+
         Ok(())
     }
 
+    /// # Safety
+    ///
+    /// The user has to assure bounds are checked.
+    #[inline]
+    pub unsafe fn set_name_unchecked(&mut self, i: usize, k: SEXP) {
```

Comment:
Is this implementation correct?
I could not figure out how to use it because `str_to_charsxp` is not public.

Reply:
Hi! It has the same implementation as `set_name`. The difference is `set_name` checks bounds. So it should work fine.
```
    pub fn set_name(&mut self, i: usize, k: &str) -> crate::error::Result<()> {
        assert_len(self.len, i)?;

        unsafe { self.set_name_unchecked(i, str_to_charsxp(k)?) };

        Ok(())
    }

    /// # Safety
    ///
    /// Calling this method with an out-of-bounds index is undefined behavior.
    #[inline]
    pub unsafe fn set_name_unchecked(&mut self, i: usize, k: SEXP) {
        if let Some(names) = self.names.as_mut() {
            unsafe { names.set_elt_unchecked(i, k) };
        }
    }
```

`set_name_unchecked` accepts a `savvy::ffi::SEXP` as argument, so if you have a `&str` you would have to create a `SEXP` as he did with `str_to_charsxp()`. Copying his function and using to create a SEXP should work fine.

```
pub(crate) unsafe fn str_to_charsxp(v: &str) -> crate::error::Result<SEXP> {
    unsafe {
        // We might be able to put `R_NaString` directly without using
        // <&str>::na(), but probably this is an inevitable cost of
        // providing <&str>::na().
        if v.is_na() {
            Ok(savvy_ffi::R_NaString)
        } else {
            crate::unwind_protect(|| {
                Rf_mkCharLenCE(
                    v.as_ptr() as *const c_char,
                    v.len() as i32,
                    cetype_t_CE_UTF8,
                )
            })
        }
    }
}
```

Reply:
I get it. Thanks for the quick answer.